### ISSUE TITLE
Feature request: relevant types in flow (i.e. flow forces you to use each property of an object at least once)

### ISSUE THREAD

**Issue by rbalicki2** [facebook/flow#4365]
Let's say I have the following situation:

* a component, which holds state of type `CustomerFormData` and which contains input elements that end up modifying the state. Assume `CustomerFormData` is a closed type, with three fields: `firstName`, `lastName`, `age`.
* a separate API function that has the signature `(CustomerFormData) => Promise<Customer>`

If age is no longer needed by the api, one can safely remove properties from `CustomerFormData`, and follow all errors (which will ultimately lead the programmer to remove the input components which affect the removed fields.) This is great! Safe refactoring!

However, if instead the API function has the signature `(CustomerPostData) => Promise<Customer>` and there is a converter of the form

```
const getCustomerPostDataFromFormData = (formData: CustomerFormData): CustomerPostData => ({
  // this is a contrived example
  fullName: `${customer.firstName} ${customer.lastName}`,
  age: customer.age,
});
```

Now, if we remove age from `CustomerPostData`, the error trail stops at `getCustomerPostDataFromFormData`! We end up with a no-op age input.

However, if there was a way to specify that in a particular context, we expect each key on a variable to be used at least once, we the error trail would not end there. Obviously, this is possible if you destructure `formData` and rely on eslint, but for larger objects that is ugly.

Would this be possible? I would love to put this on someone's radar, or even assist in implementing it, although I haven't dived into the core of the flow code base at all.

https://en.wikipedia.org/wiki/Substructural_type_system

**Comment by asolove** [comment]
If I understand you correctly, the problem you have is that returning an object with an extra `age` property typechecks even though the type doesn't expect `age` to be present and it isn't used. 

To solve that problem, you can use [exact object types](https://flow.org/en/docs/types/objects/#toc-exact-object-types). If your function is supposed to return a `type CustomerPostData = {| fullName: string |}` but it also includes an `age` property, you will get a type error.

**Comment by TiddoLangerak** [comment]
I don't think exact types will cut it here. Suppose we use exact types:

```javascript
type CustomerFormData = {|
  firstName : string,
  lastName : string,
  age : number
|};
type CustomerPostData = {|
  fullName : string, 
  age : number
|};
const getCustomerPostDataFromFormData = (formData: CustomerFormData): CustomerPostData => ({
  // this is a contrived example
  fullName: `${customer.firstName} ${customer.lastName}`,
  age: customer.age,
});
```

Now the api changes such that `age` is no longer needed. We therefore remove `age` from the `CustomerPostData`:

```javascript
type CustomerPostData = {|
  fullName : string
|};
```

We now correctly get an error in the `getCustomerPostDataFromFormData` function. We can fix this error by removing the line `age : customer.age`:

```javascript
const getCustomerPostDataFromFormData = (formData: CustomerFormData): CustomerPostData => ({
  // this is a contrived example
  fullName: `${customer.firstName} ${customer.lastName}`
});
```

Now flow won't error anymore. However, the passed in `formData` now contains a field that is no longer used. What @rbalicki2 is asking is that we could type the `formData` argument in such a way that we're forced to use each property at least once. If we could do that then flow would complain "formData.age is not used", which we could fix by either removing `age` from the type or by doing something with `age`. This in turn could let us to remove much more code that is on longer needed.

In a simple example like the above this might not seem terribly useful, since it's obvious that we can update `CustomerFormData` in parallel with `CustomerPostData`. However, in practice it might not be obvious that the fields of the 2 types are liked together, and during refactoring it is therefore very easy to miss that `CustomerFormData` can be stripped as well.

Some other examples where this would be useful (using the fictional `$Use` and `$UseProps` types to force either the value or it's properties to be used at least once).

- Require a hashing function to use both the salt and input: `hashWith(input : string, hashingFunction: ($Use<string> input, $Use<string> hash) => string)`
- Ensure that an entire object is serialized: `serialize<T>(input : $UseProps<T>) : string`
- Ensure a caller to deal with errors in a node-style callback: `doAsync(callback : (error : $Use<?mixed>, result : ?result) => mixed)`.
- Ensure the result of a promise to be used: `async function doAsync() : Promise<$Use<T>>`
- Ensure a verification function validates all properties: `function isValid(input : $UseProps<MyType>) : boolean`
